server/bin/pbench-verify-indexed.py

```diff
@@ -0,0 +1,192 @@
+                        continue
+                    if entry.name.startswith("."):
+                        # Ignore any files starting with ".".
+                        continue
+                    match = tb_pat.fullmatch(entry.name)
+                    if not match:
+                        # Ignore any directory entries which do not match the
+                        # tar ball pattern.
+                        continue
+                    yield c_entry.name, entry.name
+
+
+def main(options):
+    """Simple driver function for this CLI.
+
+    Prints a report summarizing what was found, and then provides detailed
+    lists of properly and improperly indexed tar balls.
+
+    Returns 0 on success, 1 on failure.
+    """
+    try:
+        config = pbench.PbenchConfig(options.cfg_name)
+    except pbench.BadConfig as e:
+        print(f"{_NAME_}: {e}", file=sys.stderr)
+        return 1
+
+    try:
+        archive_p = Path(config.ARCHIVE).resolve(strict=True)
+    except FileNotFoundError:
+        print(
+            f"The configured ARCHIVE directory, {config.ARCHIVE}," " does not exist",
+            file=sys.stderr,
+        )
+        return 1
+
+    if not archive_p.is_dir():
+        print(
+            f"The configured ARCHIVE directory, {config.ARCHIVE},"
+            " is not a valid directory",
+            file=sys.stderr,
+        )
+        return 1
+
+    es_client = Elasticsearch(
+        pbench.indexer._get_es_hosts(config, logging.getLogger("pbench-verify-indexed"))
+    )
+    query = {"query": {"match_all": {}}}
+
+    index_prefix = config.get("Indexing", "index_prefix")
+    if not index_prefix:
+        print(
+            "Missing 'index_prefix' value in 'Indexing' section of configuration",
+            file=sys.stderr,
+        )
+        return 1
+
+    # Start the overall timer.
+    start = pbench._time()
+    start_s = pbench.tstos(start)
+    print(f"[{start_s}] Verifying indexed state of tar balls", flush=True)
+
+    # Fetch a thousand records at a time.
+    scanner = helpers.scan(
+        es_client,
+        index=f"{index_prefix}.v4.run.*",
+        doc_type="pbench-run",
+        query=query,
+        expand_wildcards="open",
+        fields=["@metadata.file-name"],
+        size=BATCH_SIZE,
+    )
+    indexed = defaultdict(int)
+    for hit in scanner:
+        tb_file_p = Path(hit["fields"]["@metadata.file-name"][0])
+        tb_file = f"{tb_file_p.parent.name}/{tb_file_p.name}"
+        indexed[tb_file] += 1
+    indexed_cnt = len(indexed)
+
+    # Snap shot of query time.
+    q_end = pbench._time()
+
+    on_disk = defaultdict(int)
+    on_disk_cnt = 0
+    on_disk_indexed_cnt = 0
+    not_indexed = []
+    for ctrl, tb_name in gen_tb_list(archive_p):
+        tb_file = f"{ctrl}/{tb_name}"
+        on_disk[tb_file] += 1
+        on_disk_cnt += 1
+        if tb_file in indexed:
+            on_disk_indexed_cnt += 1
+        else:
+            not_indexed.append(tb_file)
```

Comment:
If `indexed` and `on_disk` were `set`'s, then you could do the `intersection()` and `difference()` to produce the `on_disk_indexed_cnt` and `not_indexed` (set) values.

Likewise for `not_on_disk`.

Reply:
I wanted the duplicate detection and it doesn't seem worth to iterate the results and not just make the checks that way.

Reply:
I'm not sure what you mean by "duplicate detection"?  (Was that something you wanted during development but subsequently removed?)

It looks to me like your code does a search of the indexed "set" for each entry in the on-disk "set"...which seems like an O(n^2) approach.  I have no idea how the `set` operations are implemented, but I would like to think that `intersection()` and `difference()` are cleverer than that.

Reply:
> I'm not sure what you mean by "duplicate detection"?

Nevermind, I see, now.

I'll just point out that you can still do duplicate detection if you use a `set`-based approach, but it does take some of the beauty out of it.